controller/src/main/java/org/jboss/as/controller/ControlledProcessState.java

```diff
@@ -117,6 +120,7 @@ public void setStarting() {
 
     public void setRunning() {
         AtomicStampedReference<State> stateRef = state;
+        int currentStamp = stamp.get();
         int newStamp = stamp.incrementAndGet();
```

Comment:
I just noticed a race here. I think this should be:

int currentStamp = stamp.getAndIncrement();
int newStamp = currentStamp + 1; // the value from the increment we just did is "ours"

Reply:
If we increment it, we won't be able to revert from a restart/reload required. The issue is the stamp saved in the step that executed the setRestartRequired()/setReloadRequired() does not change, because it is an artificial state derived from a flag variable

Reply:
If you want to see it in a different way, the `stamp.getAndIncrement()` you are requesting to do here, was already done when setRestartRequired()/setReloadRequired() was called: [ControlledProcessState.java#L192](https://github.com/wildfly/wildfly-core/blob/757e3012d9a14e3090fc129c22a82b15514d3e58/controller/src/main/java/org/jboss/as/controller/ControlledProcessState.java#L192)